File: sdk/python/alp_sdk/domain_trust.py
"""ALP Cross-Domain Trust (v18.4.0 — V14 The Sovereign Era).

Trust bootstrapping between sovereign domains:

* ``DomainTrustAnchor`` — exchanges signed trust roots with foreign domains.
* ``TrustRoot``        — signed root of trust for a domain.
* ``DomainLink``       — bilateral trust relationship between domains.

Enables cross-domain agent authentication without a global CA.
``alp trust link <domain>`` establishes bilateral trust.

Mirrors the planned ``parser/src/domain_trust.ts`` surface; tests live in
``sdk/python/tests/test_domain_trust.py``.
"""

import hashlib
import json
import os
import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional, Set


TRUST_DIR = ".trust"
DOMAINS_FILE = "domains.jsonl"
LINKS_FILE = "links.jsonl"


def trust_dir(alp_dir: str) -> str:
    return os.path.join(alp_dir, TRUST_DIR)


def domains_path(alp_dir: str) -> str:
    return os.path.join(trust_dir(alp_dir), DOMAINS_FILE)


def links_path(alp_dir: str) -> str:
    return os.path.join(trust_dir(alp_dir), LINKS_FILE)


class TrustStatus(str, Enum):
    ACTIVE = "active"
    PENDING = "pending"
    REVOKED = "revoked"
    EXPIRED = "expired"


@dataclass
class TrustRoot:
    domain_id: str
    public_key: str
    signature: str = ""
    created_at: str = ""
    expires_at: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        if not self.created_at:
            self.created_at = _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "domain_id": self.domain_id,
            "public_key": self.public_key,
            "signature": self.signature,
            "created_at": self.created_at,
            "expires_at": self.expires_at,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "TrustRoot":
        return cls(
            domain_id=d["domain_id"],
            public_key=d["public_key"],
            signature=d["signature"],
            created_at=d.get("created_at", ""),
            expires_at=d.get("expires_at"),
            metadata=d.get("metadata", {}),
        )

    def sign(self, private_key: str) -> None:
        payload = json.dumps({
            "domain_id": self.domain_id,
            "public_key": self.public_key,
            "created_at": self.created_at,
        }, sort_keys=True).encode()
        self.signature = hashlib.sha256(payload + private_key.encode()).hexdigest()

    def verify(self, private_key: str) -> bool:
        payload = json.dumps({
            "domain_id": self.domain_id,
            "public_key": self.public_key,
            "created_at": self.created_at,
        }, sort_keys=True).encode()
        expected = hashlib.sha256(payload + private_key.encode()).hexdigest()
        return self.signature == expected


@dataclass
class DomainLink:
    link_id: str
    local_domain: str
    remote_domain: str
    status: str = "pending"
    created_at: str = ""
    accepted_at: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        if not self.created_at:
            self.created_at = _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "link_id": self.link_id,
            "local_domain": self.local_domain,
            "remote_domain": self.remote_domain,
            "status": self.status,
            "created_at": self.created_at,
            "accepted_at": self.accepted_at,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "DomainLink":
        return cls(
            link_id=d["link_id"],
            local_domain=d["local_domain"],
            remote_domain=d["remote_domain"],
            status=d.get("status", "pending"),
            created_at=d.get("created_at", ""),
            accepted_at=d.get("accepted_at", ""),
            metadata=d.get("metadata", {}),
        )


@dataclass
class DomainTrustAnchor:
    alp_dir: str
    domain_id: str
    private_key: str

    def __post_init__(self):
        self._domain_dir = os.path.join(trust_dir(self.alp_dir), "domains", self.domain_id)
        self._root_path = os.path.join(self._domain_dir, "root.json")

    def create_domain(self, metadata: Optional[Dict[str, Any]] = None) -> TrustRoot:
        public_key = hashlib.sha256(self.private_key.encode()).hexdigest()
        root = TrustRoot(domain_id=self.domain_id, public_key=public_key, signature="", metadata=metadata or {})
        root.sign(self.private_key)
        d = os.path.dirname(self._root_path)
        if not os.path.exists(d):
            os.makedirs(d, exist_ok=True)
        with open(self._root_path, "w", encoding="utf-8") as f:
            json.dump(root.to_dict(), f, indent=2)
        return root

    def get_trust_root(self) -> Optional[TrustRoot]:
        if not os.path.exists(self._root_path):
            return None
        with open(self._root_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        return TrustRoot.from_dict(data)

    def export_trust_root(self) -> str:
        root = self.get_trust_root()
        if not root:
            raise ValueError(f"No trust root found for domain '{self.domain_id}'.")
        return json.dumps(root.to_dict(), indent=2)

    def import_trust_root(self, remote_root_json: str, expected_domain_id: str) -> TrustRoot:
        data = json.loads(remote_root_json)
        root = TrustRoot.from_dict(data)
        if root.domain_id != expected_domain_id:
            raise ValueError(f"Domain ID mismatch: expected '{expected_domain_id}', got '{root.domain_id}'.")
        if not root.verify(self.private_key):
            raise ValueError("Trust root signature verification failed.")
        return root


class DomainTrustManager:
    """Manages trust links between local and remote domains."""

    def __init__(self, alp_dir: str, local_domain: str):
        self.alp_dir = alp_dir
        self.local_domain = local_domain
        self._links: Dict[str, DomainLink] = {}
        self._load()

    def _load(self) -> None:
        p = links_path(self.alp_dir)
        if not os.path.exists(p):
            return
        try:
            with open(p, "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    entry = json.loads(line)
                    link = DomainLink.from_dict(entry)
                    self._links[link.link_id] = link
        except Exception:
            self._links = {}

    def _save_link(self, link: DomainLink) -> None:
        d = trust_dir(self.alp_dir)
        if not os.path.exists(d):
            os.makedirs(d, exist_ok=True)
        with open(links_path(self.alp_dir), "a", encoding="utf-8") as f:
            f.write(json.dumps(link.to_dict()) + "\n")

    def link_domain(self, remote_domain: str) -> DomainLink:
        existing = [l for l in self._links.values() if l.remote_domain == remote_domain]
        if existing:
            return existing[0]
        link_id = f"link-{uuid.uuid4().hex[:12]}"
        link = DomainLink(link_id=link_id, local_domain=self.local_domain, remote_domain=remote_domain, status="pending")
        self._links[link_id] = link
        self._save_link(link)
        return link

    def accept_link(self, link_id: str) -> Optional[DomainLink]:
        link = self._links.get(link_id)
        if not link:
            return None
        link.status = "active"
        link.accepted_at = _now_iso()
        self._save_link(link)
        return link

    def revoke_link(self, link_id: str) -> bool:
        link = self._links.get(link_id)
        if not link:
            return False
        link.status = "revoked"
        self._save_link(link)
        return True

    def get_link(self, link_id: str) -> Optional[DomainLink]:
        return self._links.get(link_id)

    def get_link_by_domain(self, remote_domain: str) -> Optional[DomainLink]:
        for link in self._links.values():
            if link.remote_domain == remote_domain:
                return link
        return None

    def list_links(self) -> List[DomainLink]:
        return list(self._links.values())

    def list_active_links(self) -> List[DomainLink]:
        return [l for l in self._links.values() if l.status == "active"]


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def create_domain_keypair() -> tuple[str, str]:
    private_key = os.urandom(32).hex()
    public_key = hashlib.sha256(private_key.encode()).hexdigest()
    return public_key, private_key

File: parser/src/domain_trust.ts
/** ALP Cross-Domain Trust (v18.4.0 — V14 The Sovereign Era).
 *
 * Trust bootstrapping between sovereign domains:
 *
 * - `TrustRoot`        — signed root of trust for a domain.
 * - `DomainLink`       — bilateral trust relationship between domains.
 * - `DomainTrustAnchor` — exchanges signed trust roots with foreign domains.
 * - `DomainTrustManager` — manages trust links between local and remote domains.
 *
 * Enables cross-domain agent authentication without a global CA.
 * `alp trust link <domain>` establishes bilateral trust.
 *
 * Mirrors `sdk/python/alp_sdk/domain_trust.py`.
 */

import * as crypto from 'node:crypto'
import * as fs from 'fs'
import * as path from 'path'

export const TRUST_DIR = '.trust'
export const DOMAINS_FILE = 'domains.jsonl'
export const LINKS_FILE = 'links.jsonl'

export enum TrustStatus {
  ACTIVE = 'active',
  PENDING = 'pending',
  REVOKED = 'revoked',
  EXPIRED = 'expired',
}

export interface TrustRootData {
  domain_id: string
  public_key: string
  signature: string
  created_at: string
  expires_at: string | null
  metadata: Record<string, any>
}

export class TrustRoot {
  domain_id: string
  public_key: string
  signature: string
  created_at: string
  expires_at: string | null
  metadata: Record<string, any>

  constructor(domain_id: string, public_key: string, signature = '', created_at = '', expires_at: string | null = null, metadata: Record<string, any> = {}) {
    this.domain_id = domain_id
    this.public_key = public_key
    this.signature = signature
    this.created_at = created_at || new Date().toISOString()
    this.expires_at = expires_at
    this.metadata = metadata
  }

  toDict(): TrustRootData {
    return {
      domain_id: this.domain_id,
      public_key: this.public_key,
      signature: this.signature,
      created_at: this.created_at,
      expires_at: this.expires_at,
      metadata: this.metadata,
    }
  }

  static fromDict(d: Record<string, any>): TrustRoot {
    return new TrustRoot(
      d.domain_id ?? d['domain_id'],
      d.public_key ?? d['public_key'],
      d.signature ?? d['signature'],
      d.created_at ?? '',
      d.expires_at ?? null,
      d.metadata ?? {},
    )
  }

  sign(private_key: string): void {
    const payload_obj = {
      domain_id: this.domain_id,
      public_key: this.public_key,
      created_at: this.created_at,
    }
    const payload = JSON.stringify(payload_obj, Object.keys(payload_obj).sort())
    this.signature = crypto.createHash('sha256').update(payload + private_key).digest('hex')
  }

  verify(private_key: string): boolean {
    const payload_obj = {
      domain_id: this.domain_id,
      public_key: this.public_key,
      created_at: this.created_at,
    }
    const payload = JSON.stringify(payload_obj, Object.keys(payload_obj).sort())
    const expected = crypto.createHash('sha256').update(payload + private_key).digest('hex')
    return this.signature === expected
  }
}

export interface DomainLinkData {
  link_id: string
  local_domain: string
  remote_domain: string
  status: string
  created_at: string
  accepted_at: string
  metadata: Record<string, any>
}

export class DomainLink {
  link_id: string
  local_domain: string
  remote_domain: string
  status: string
  created_at: string
  accepted_at: string
  metadata: Record<string, any>

  constructor(link_id: string, local_domain: string, remote_domain: string, status = 'pending', created_at = '', accepted_at = '', metadata: Record<string, any> = {}) {
    this.link_id = link_id
    this.local_domain = local_domain
    this.remote_domain = remote_domain
    this.status = status
    this.created_at = created_at || new Date().toISOString()
    this.accepted_at = accepted_at
    this.metadata = metadata
  }

  toDict(): DomainLinkData {
    return {
      link_id: this.link_id,
      local_domain: this.local_domain,
      remote_domain: this.remote_domain,
      status: this.status,
      created_at: this.created_at,
      accepted_at: this.accepted_at,
      metadata: this.metadata,
    }
  }

  static fromDict(d: Record<string, any>): DomainLink {
    return new DomainLink(
      d.link_id ?? d['link_id'],
      d.local_domain ?? d['local_domain'],
      d.remote_domain ?? d['remote_domain'],
      d.status ?? 'pending',
      d.created_at ?? '',
      d.accepted_at ?? '',
      d.metadata ?? {},
    )
  }
}

export class DomainTrustAnchor {
  private alp_dir: string
  private domain_id: string
  private private_key: string
  private _domain_dir: string
  private _root_path: string

  constructor(alp_dir: string, domain_id: string, private_key: string) {
    this.alp_dir = alp_dir
    this.domain_id = domain_id
    this.private_key = private_key
    this._domain_dir = path.join(this.alp_dir, TRUST_DIR, 'domains', this.domain_id)
    this._root_path = path.join(this._domain_dir, 'root.json')
  }

  create_domain(metadata: Record<string, any> = {}): TrustRoot {
    const public_key = crypto.createHash('sha256').update(this.private_key).digest('hex')
    const root = new TrustRoot(this.domain_id, public_key, '', '', null, metadata)
    root.sign(this.private_key)
    const d = path.dirname(this._root_path)
    if (!fs.existsSync(d)) fs.mkdirSync(d, { recursive: true })
    fs.writeFileSync(this._root_path, JSON.stringify(root.toDict(), null, 2))
    return root
  }

  get_trust_root(): TrustRoot | null {
    if (!fs.existsSync(this._root_path)) return null
    const data = JSON.parse(fs.readFileSync(this._root_path, 'utf-8'))
    return TrustRoot.fromDict(data)
  }

  export_trust_root(): string {
    const root = this.get_trust_root()
    if (!root) throw new Error(`No trust root found for domain '${this.domain_id}'.`)
    return JSON.stringify(root.toDict(), null, 2)
  }

  import_trust_root(remote_root_json: string, expected_domain_id: string): TrustRoot {
    const data = JSON.parse(remote_root_json)
    const root = TrustRoot.fromDict(data)
    if (root.domain_id !== expected_domain_id) {
      throw new Error(`Domain ID mismatch: expected '${expected_domain_id}', got '${root.domain_id}'.`)
    }
    if (!root.verify(this.private_key)) {
      throw new Error('Trust root signature verification failed.')
    }
    return root
  }
}

export class DomainTrustManager {
  private alp_dir: string
  private local_domain: string
  private links: Map<string, DomainLink> = new Map()

  constructor(alp_dir: string, local_domain: string) {
    this.alp_dir = alp_dir
    this.local_domain = local_domain
    this.load()
  }

  private links_path(): string {
    return path.join(this.alp_dir, TRUST_DIR, LINKS_FILE)
  }

  load(): void {
    const p = this.links_path()
    if (!fs.existsSync(p)) return
    try {
      const lines = fs.readFileSync(p, 'utf-8').split('\n').filter(Boolean)
      for (const line of lines) {
        const entry = JSON.parse(line)
        const link = DomainLink.fromDict(entry)
        this.links.set(link.link_id, link)
      }
    } catch {
      this.links.clear()
    }
  }

  private save_link(link: DomainLink): void {
    const d = path.join(this.alp_dir, TRUST_DIR)
    if (!fs.existsSync(d)) fs.mkdirSync(d, { recursive: true })
    fs.appendFileSync(this.links_path(), JSON.stringify(link.toDict()) + '\n')
  }

  link_domain(remote_domain: string): DomainLink {
    const existing = Array.from(this.links.values()).filter((l) => l.remote_domain === remote_domain)
    if (existing.length > 0) return existing[0]
    const link_id = `link-${crypto.randomBytes(6).toString('hex')}`
    const link = new DomainLink(link_id, this.local_domain, remote_domain, 'pending')
    this.links.set(link_id, link)
    this.save_link(link)
    return link
  }

  accept_link(link_id: string): DomainLink | null {
    const link = this.links.get(link_id)
    if (!link) return null
    link.status = TrustStatus.ACTIVE
    link.accepted_at = new Date().toISOString()
    this.save_link(link)
    return link
  }

  revoke_link(link_id: string): boolean {
    const link = this.links.get(link_id)
    if (!link) return false
    link.status = TrustStatus.REVOKED
    this.save_link(link)
    return true
  }

  get_link(link_id: string): DomainLink | undefined {
    return this.links.get(link_id)
  }

  get_link_by_domain(remote_domain: string): DomainLink | undefined {
    return Array.from(this.links.values()).find((l) => l.remote_domain === remote_domain)
  }

  list_links(): DomainLink[] {
    return Array.from(this.links.values())
  }

  list_active_links(): DomainLink[] {
    return Array.from(this.links.values()).filter((l) => l.status === TrustStatus.ACTIVE)
  }
}

export function create_domain_keypair(): { public_key: string; private_key: string } {
  const private_key = crypto.randomBytes(32).toString('hex')
  const public_key = crypto.createHash('sha256').update(private_key).digest('hex')
  return { public_key, private_key }
}
